pkg/catalogsourceconfig/registry.go

```diff
@@ -0,0 +1,369 @@
+		r.log.Infof("Created Deployment %s with registry command: %s", deployment.GetName(), registryCommand)
+	} else {
+		// Update the pod specification
+		deployment.Spec.Template = r.newPodTemplateSpec(registryCommand)
+		err = r.client.Update(context.TODO(), deployment)
+		if err != nil {
+			r.log.Errorf("Failed to update Deployment %s : %v", deployment.GetName(), err)
+			return err
+		}
+		r.log.Infof("Updated Deployment %s with registry command: %s", deployment.GetName(), registryCommand)
+	}
+	return nil
+}
+
+// ensureRole ensure that the Role required to access the given operatorSources
+// from the registry Deployment is present.
+func (r *registry) ensureRole(operatorSources []string) error {
+	role := new(RoleBuilder).WithTypeMeta().Role()
+	if err := r.client.Get(context.TODO(), r.csc.key(), role); err != nil {
+		role = r.newRole(operatorSources)
+		err = r.client.Create(context.TODO(), role)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			r.log.Errorf("Failed to create Role %s: %v", role.GetName(), err)
+			return err
+		}
+		r.log.Infof("Created Role %s", role.GetName())
+	} else {
+		// Update the Rules to be on the safe side
+		role.Rules = getRules(operatorSources)
+		err = r.client.Update(context.TODO(), role)
+		if err != nil {
+			r.log.Errorf("Failed to update Role %s : %v", role.GetName(), err)
+			return err
+		}
+		r.log.Infof("Updated Role %s", role.GetName())
+	}
+	return nil
+}
+
+// ensureRoleBinding ensures that the RoleBinding bound to the Role previously
+// created is present.
+func (r *registry) ensureRoleBinding() error {
+	roleBinding := new(RoleBindingBuilder).WithTypeMeta().RoleBinding()
+	if err := r.client.Get(context.TODO(), r.csc.key(), roleBinding); err != nil {
+		roleBinding = r.newRoleBinding(r.csc.GetName())
+		err = r.client.Create(context.TODO(), roleBinding)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			r.log.Errorf("Failed to create RoleBinding %s: %v", roleBinding.GetName(), err)
+			return err
+		}
+		r.log.Infof("Created RoleBinding %s", roleBinding.GetName())
+	} else {
+		// Update the Rules to be on the safe side
+		roleBinding.RoleRef = NewRoleRef(r.csc.GetName())
+		err = r.client.Update(context.TODO(), roleBinding)
+		if err != nil {
+			r.log.Errorf("Failed to update RoleBinding %s : %v", roleBinding.GetName(), err)
+			return err
+		}
+		r.log.Infof("Updated RoleBinding %s", roleBinding.GetName())
+	}
+	return nil
+}
+
+// ensureService ensure that the Service for the registry deployment is present.
+func (r *registry) ensureService() error {
+	service := new(ServiceBuilder).WithTypeMeta().Service()
+	// Delete the Service so that we get a new ClusterIP
+	if err := r.client.Get(context.TODO(), r.csc.key(), service); err == nil {
+		r.log.Infof("Service %s is present", service.GetName())
+		err := r.client.Delete(context.TODO(), service)
+		if err != nil {
+			r.log.Errorf("Failed to delete Service %s", service.GetName())
+			// Make a best effort to create the service
```

Comment:
If it can't delete, then the logs will say `failed to delete` and then say `deleted the object`.

Reply:
I will fix this